core/key.go

```diff
@@ -472,52 +484,116 @@ func (k DataStoreKey) ToPrimaryDataStoreKey() PrimaryDataStoreKey {
 //
-// /[CollectionID]/[IndexID]/[FieldValue](/[FieldValue]...)
+// /[CollectionID]/[IndexID]/[FieldID][FieldKind][FieldValue](/[FieldID][FieldKind][FieldValue]...)
 //
-// Where [CollectionID] and [IndexID] are integers
-func NewIndexDataStoreKey(key string) (IndexDataStoreKey, error) {
-	if key == "" {
+// Where [CollectionID], [IndexID], [FieldID] and [FieldKind] are integers
+func DecodeIndexDataStoreKey(b []byte, indexDesc *client.IndexDescription) (IndexDataStoreKey, error) {
+	if len(b) == 0 {
 		return IndexDataStoreKey{}, ErrEmptyKey
 	}
 
-	if !strings.HasPrefix(key, "/") {
+	key := IndexDataStoreKey{}
+
+	if b[0] != '/' {
 		return IndexDataStoreKey{}, ErrInvalidKey
 	}
+	b = b[1:]
+	b, colID, err := encoding.DecodeUvarintAscending(b)
+	if err != nil {
+		return IndexDataStoreKey{}, err
+	}
+	key.CollectionID = uint32(colID)
 
-	elements := strings.Split(key[1:], "/")
-
-	// With less than 3 elements, we know it's an invalid key
-	if len(elements) < 3 {
+	if b[0] != '/' {
 		return IndexDataStoreKey{}, ErrInvalidKey
 	}
-
-	colID, err := strconv.Atoi(elements[0])
+	b = b[1:]
+	b, indID, err := encoding.DecodeUvarintAscending(b)
 	if err != nil {
-		return IndexDataStoreKey{}, ErrInvalidKey
+		return IndexDataStoreKey{}, err
 	}
+	key.IndexID = uint32(indID)
 
-	indexKey := IndexDataStoreKey{CollectionID: uint32(colID)}
-
-	indID, err := strconv.Atoi(elements[1])
-	if err != nil {
-		return IndexDataStoreKey{}, ErrInvalidKey
+	if len(b) == 0 {
+		return key, nil
 	}
-	indexKey.IndexID = uint32(indID)
 
-	// first 2 elements are the collection and index IDs, the rest are field values
-	for i := 2; i < len(elements); i++ {
-		indexKey.FieldValues = append(indexKey.FieldValues, []byte(elements[i]))
+	for len(b) > 0 {
+		var fieldID, fieldKind uint64
+		if b[0] != '/' {
+			return IndexDataStoreKey{}, ErrInvalidKey
+		}
+		b = b[1:]
+		b, fieldID, err = encoding.DecodeUvarintAscending(b)
+		if err != nil {
+			return IndexDataStoreKey{}, err
+		}
+		b, fieldKind, err = encoding.DecodeUvarintAscending(b)
+		if err != nil {
+			return IndexDataStoreKey{}, err
+		}
+
+		var fieldVal *client.FieldValue
+		i := len(key.Fields)
+		descending := false
+		if i < len(indexDesc.Fields) { // TOTEST
+			descending = indexDesc.Fields[i].Descending
+		}
+		b, fieldVal, err = encoding.DecodeFieldValue(b, client.FieldKind(fieldKind), descending)
+		if err != nil {
+			return IndexDataStoreKey{}, err
+		}
+
+		key.Fields = append(key.Fields, IndexedField{
+			ID:    client.FieldID(fieldID),
+			Value: fieldVal,
+		})
 	}
 
-	return indexKey, nil
+	return key, nil
 }
 
 // Bytes returns the byte representation of the key
 func (k *IndexDataStoreKey) Bytes() []byte {
-	return []byte(k.ToString())
+	b, _ := EncodeIndexDataStoreKey([]byte{}, k)
+	return b
+}
+
+// EncodeIndexDataStoreKey encodes a IndexDataStoreKey to bytes to be stored as a key
+// for secondary indexes.
+func EncodeIndexDataStoreKey(b []byte, key *IndexDataStoreKey) ([]byte, error) {
```

Comment:
what do you mean "in the constructor"?

Reply:
The function that creates `IndexDataStoreKey` instances

Reply:
As I undestood, `Bytes() []byte` has this signature to match some interface from `ipsf` (I guess `ipfs.Key`). In this case changing it is not an option.

Making a constructor won't solve the problem of trying to make it like "the rest of serialisation functions" because all the key-structs in this file are public and in order to be sure encoding won't fail, we have to either encode on constructon, which will be wastefull for resources, or do some kind of validation to make sure encoding will work. In both of these cases we will have to make the struct private so that nobody can tinker with.

I didn't think of any other solution because my brain is busy with other part of code at the moment. Just dropping a comment in case you have a more specific suggestion.

Reply:
> Bytes() []byte has this signature to match some interface from ipsf (I guess ipfs.Key)

I am fairly confident that is not the case, and if it was changed to do so the `ipfs.Key` interface should have been embedded in the `core.Key` interface (which it is not).

> In both of these cases we will have to make the struct private so that nobody can tinker with

Whilst not ideal, I personally am okay with not worry about that for internal types - is just a weakness of Go lack of access modifiers.  IMO the constructors should be used all of the time anyway (if/when they are provided), if nothing else in a language that doesn't compiler-enforce the setting of all the props on init, it makes it much much easier to track all the places where it is initialised.

The waste of resources is not wonderful though, but these are never created in very performance sensitive locations no (not like doc-keys)? Although coupled with Go's weaknesses I think I'm leaning more towards changing the `Bytes` signature.

Changing `Bytes` would be a bit of a hassle, but very probably much less hassle than silently dropping errors.

Reply:
> is just a weakness of Go lack of access modifiers

It has access modifiers. You can have public, private and internal public :)

Reply:
*is just a weakness of Go lacking many useful access modifiers found in many other programming languages

Reply:
> IMO the constructors should be used all of the time anyway (if/when they are provided), if nothing else in a language that doesn't compiler-enforce the setting of all the props on init

somewhat off topic but, I have had similar thoughts, and wonder if long-term this would be helpful: https://github.com/GaijinEntertainment/go-exhaustruct

Reply:
That's quite cool, I never thought there might be a linter like thing for that in Go.

I suspect we have quite a lot of code that is dependent on this behaviour though, and I'm not sure everyone would be onboard with the principle.  Could be a fun discussion in the cooldown though, and something I might try in ~~the lens repo~~hobby projects.

Reply:
I'd be 100% onboard with this. If i recall perhaps i even might have had opened an issue for this linter. This was when a while back we saw some filterAggregate bug seap through due to filter not being set at some init IIRC. Can be a nice addition

Reply:
refactored